pkg/storage/stores/tsdb/index_shipper_querier.go

```diff
@@ -29,36 +31,35 @@ func newIndexShipperQuerier(shipper indexShipperIterator, tableRanges config.Tab
 	return &indexShipperQuerier{shipper: shipper, tableRanges: tableRanges}
 }
 
+type indexIterFunc func(func(context.Context, Index) error) error
+
+func (i indexIterFunc) For(ctx context.Context, f func(context.Context, Index) error) error {
+	return i(f)
+}
+
 func (i *indexShipperQuerier) indices(ctx context.Context, from, through model.Time, user string) (Index, error) {
-	var indices []Index
-
-	// Ensure we query both per tenant and multitenant TSDBs
-	idxBuckets := indexBuckets(from, through, i.tableRanges)
-	for _, bkt := range idxBuckets {
-		if err := i.shipper.ForEach(ctx, bkt, user, func(multitenant bool, idx shipper_index.Index) error {
-			impl, ok := idx.(Index)
-			if !ok {
-				return fmt.Errorf("unexpected shipper index type: %T", idx)
+	itr := indexIterFunc(func(f func(context.Context, Index) error) error {
+		// Ensure we query both per tenant and multitenant TSDBs
+		idxBuckets := indexBuckets(from, through, i.tableRanges)
+		for _, bkt := range idxBuckets {
+			if err := i.shipper.ForEachConcurrent(ctx, bkt, user, func(multitenant bool, idx shipper_index.Index) error {
+				impl, ok := idx.(Index)
+				if !ok {
+					return fmt.Errorf("unexpected shipper index type: %T", idx)
+				}
+				if multitenant {
+					impl = NewMultiTenantIndex(impl)
+				}
+
+				return f(ctx, impl)
+			}); err != nil {
+				return err
```

Comment:
i think here you're allowing `itr` to be of type `err`. I'm not sure the NewMultiIndex at line 62 will be able of handling it though :thinking:

Reply:
The return here is part of the function definition cast to `indexIterFunc`. It looks fine to me. `itr` is holding a function definition and not capturing the output of a function call.

Reply:
ohhh you right, thank you